### ISSUE TITLE
Global state mocking

### ISSUE THREAD

**Issue by egeerardyn** [matlab2tikz/matlab2tikz#542]
We should write some code to mock in global state, e.g.
- the DPI of the screen used
- the values in `get(0, 'Default')` and `get(0, 'Factory')`
- maybe some others

E.g. a thin wrapper around the standard mechanisms to be able to inject our own state instead of the global state.

Especially for `get(0, 'Default')` etc. we might also consider having a mechanism to retrieve all those settings, set them to empty and reset to the retrieved values. This would considerably increase testability and the reliability of our test cases (see also #485), certainly over different machines and environments.

**Comment by PeterPablo** [comment]
There are several possible courses of action. We can go for a brute force approach and setting/resetting the complete default values or take more care and modify only specific parts of the default graphical behavior.

In #552 I wrote a small snippet that counters the `\definecolor` statements in R2014b output. I agree on @egeerardyn opinion that unifying all output is very hard and that we shouldn't spend too much time. But I think that there is nothing wrong with picking some *low hanging fruits*.

defaultAxesColorOrder = [0,0,1;0,0.500,0;1,0,0;0,0.750,0.750;0.750,0,0.750;0.750,0.750,0;0.250,0.250,0.250;]; set(0,'defaultAxesColorOrder',defaultAxesColorOrder);

**Comment by okomarov** [comment]
We could just blindly save the Octave state used in Travis into a .mat file and that's it. We only need to make sure in `m2t` that the DPI is always retrieved from `get(0)`. 

Setting default values to empty might cause unexpected behaviour.

**Comment by egeerardyn** [comment]
The `get(0)` values such as DPI might change from dev machine to dev machine. So always getting it from there is one of the causes of unreliable tests as we've witnessed in #485. That's why we'd need to be able to mock those values inside `m2t` (not only testing, but the actual production code): such that we can set it to a value we like, independent of the environment we're in.

With "empty default values", I meant to have the distributed default values (i.e. factory values). This may indeed provide inconsistent results as they might change over time. In those cases, we should provide our own values.

**Comment by okomarov** [comment]
I suggest to approach the global state mocking in two separate steps: one that solves testing issues and the other more general for production code. The former can solve the hashes problem quite quickly and should be easy to implement. Then we can work on the default state for `m2t` which requires a selective approach that is likely to take more time to implement.

**Comment by egeerardyn** [comment]
@okomarov: I agree completely with that approach!

First fix the symptoms of the problem in testing (get the global state, reset to something known, run the tests, reset the global state to user state). That will make most of the tests more reliable.

Afterwards, we should actually mock the state in production to make testing easier to make the rest of the tests (e.g. DPI-dependent ones) reliable as well.

**Comment by egeerardyn** [comment]
As requested in #639: my global state in R2014a
```
>> get(0)
	CallbackObject = []
	CommandWindowSize = [175 23]
	CurrentFigure = [4]
	Diary = off
	DiaryFile = diary
	Echo = off
	FixedWidthFontName = Courier New
	Format = shortG
	FormatSpacing = compact
	Language = en_us
	MonitorPositions = [0 0 1920 1200]
	More = off
	PointerLocation = [1015 209]
	RecursionLimit = [500]
	ScreenDepth = [24]
	ScreenPixelsPerInch = [72.0567]
	ScreenSize = [1 1 1920 1200]
	ShowHiddenHandles = off
	Units = pixels

	BeingDeleted = off
	ButtonDownFcn = 
	Children = [ (4 by 1) double array]
	Clipping = on
	CreateFcn = 
	DeleteFcn = 
	BusyAction = queue
	HandleVisibility = on
	HitTest = on
	Interruptible = on
	Parent = []
	Selected = off
	SelectionHighlight = on
	Tag = 
	Type = root
	UIContextMenu = []
	UserData = []
	Visible = on

>> get(0,'Default')
ans = 
             defaultFigurePosition: [680 678 560 420]
                  defaultTextColor: [0 0 0]
                 defaultAxesXColor: [0 0 0]
                 defaultAxesYColor: [0 0 0]
                 defaultAxesZColor: [0 0 0]
             defaultPatchFaceColor: [0 0 0]
             defaultPatchEdgeColor: [0 0 0]
                  defaultLineColor: [0 0 0]
       defaultFigureInvertHardcopy: 'on'
                  defaultAxesColor: [1 1 1]
             defaultAxesColorOrder: [7x3 double]
             defaultFigureColormap: [64x3 double]
           defaultSurfaceEdgeColor: [0 0 0]
                defaultFigureColor: [1 1 1]
            defaultFigurePaperType: 'A4'
    defaultFigurePaperPositionMode: 'auto'
          defaultUicontrolFontSize: 10
```
I will edit this post to add R2014b values when I start up that version.

**Comment by PeterPablo** [comment]
## My R2014a (Linux):
```
>> get(0)
	CallbackObject = []
	CommandWindowSize = [137 20]
	CurrentFigure = []
	Diary = off
	DiaryFile = diary
	Echo = off
	FixedWidthFontName = Courier New
	Format = short
	FormatSpacing = loose
	Language = de_de
	MonitorPositions = [0 0 3200 1200]
	More = off
	PointerLocation = [2393 320]
	RecursionLimit = [500]
	ScreenDepth = [24]
	ScreenPixelsPerInch = [90]
	ScreenSize = [1 1 3200 1200]
	ShowHiddenHandles = off
	Units = pixels

	BeingDeleted = off
	ButtonDownFcn = 
	Children = []
	Clipping = on
	CreateFcn = 
	DeleteFcn = 
	BusyAction = queue
	HandleVisibility = on
	HitTest = on
	Interruptible = on
	Parent = []
	Selected = off
	SelectionHighlight = on
	Tag = 
	Type = root
	UIContextMenu = []
	UserData = []
	Visible = on
```

## My R2014b (Linux):
```
>> get(0)
         CallbackObject: []
               Children: []
          CurrentFigure: []
     FixedWidthFontName: 'Courier New'
       HandleVisibility: 'on'
       MonitorPositions: [2x4 double]
                 Parent: []
        PointerLocation: [1790 180]
            ScreenDepth: 24
    ScreenPixelsPerInch: 95
             ScreenSize: [1 1 1600 1200]
      ShowHiddenHandles: 'off'
                    Tag: ''
                   Type: 'root'
                  Units: 'pixels'
               UserData: []
```

## Conclusion from issue #639
Applying, set(0,'ScreenPixelsPerInch',72.0567), as suggested by your configuration (#542 (comment)) makes the output of ACID(69) identical! :+1:

This is a strong indication, that we should follow this approach in #542 to solve #552.
I propse to start in a first step with adapting only this setting and extend this approach to more properties, if this is even necessary.

**Comment by PeterPablo** [comment]
@egeerardyn, your R2014b output would still be appreciated.

**Interesting properties for us:**
* `defaultFigurePosition = [680 678 560 420]` : especially `width` and `height` (equal on reported output)
* `defaultFigureColor` differs between the two reported outputs on R2014a

**Comment by egeerardyn** [comment]
R2014b output:
```
>> get(0)
         CallbackObject: []
               Children: []
          CurrentFigure: []
     FixedWidthFontName: 'Courier New'
       HandleVisibility: 'on'
       MonitorPositions: [1 1 1920 1200]
                 Parent: []
        PointerLocation: [797 314]
            ScreenDepth: 32
    ScreenPixelsPerInch: 72
             ScreenSize: [1 1 1920 1200]
      ShowHiddenHandles: 'off'
                    Tag: ''
                   Type: 'root'
                  Units: 'pixels'
               UserData: []
>> get(0,'Default')
ans = 
    defaultFigurePaperPositionMode: 'auto'
             defaultFigurePosition: [680 678 560 420]
                defaultFigureColor: [1 1 1]
            defaultFigurePaperType: 'A4'
```

**Comment by PeterPablo** [comment]
Thank you. No new surprises. I extended #641 and it now also covers `ScreenDepth`, although I have the impression this does not influence the output.

**Comment by egeerardyn** [comment]
@PeterPablo I think that `ScreenDepth` should only influence the rendering on-screen, not the actual data structures we are iterating over.

**Comment by PeterPablo** [comment]
Reading [the documentation](http://de.mathworks.com/help/images/displaying-colors.html) I have the impression that especially 8 bit might influence the resulting output, if m2t can't match the colors to known literals. I didn't try this myself.

**Comment by PeterPablo** [comment]
Implemented by #641. Closing here.
